File: intro/c/primer.c
/* Learning Objectives

- Reason for learning C:
-- The Python language hides from the user the many stages taken by the interpreter software to translate the Python code into a machine-level encoding.

- 

*/ 

// start translation:
/*
- Like the Python language defines a print function,
- C defines a printf function.
- Unlike the Python language, in C programming, every pre-defined function must be credited to its source file - i.e. its implementation.
- In C, #include<filename> is used to import code not in the current file for use by the current file.

- printf does not automatically add a new line like a print function.
- Hence, we have to type in the character encoding of a new line.
- In the language of programming - i.e. ASCII - the character encoding for a new line is /n.

- The C compiler will take care of converting the characters in the ASCII string to binary code/language.

- The C compiler requires that every single line of code is ended with a ; by the programmer. This informs the compiler of what a single line of code is.
- Unlike Python, C is compiled in full, not line by line. Python by its design separates its scripts into lines.

*/

/*
#include<stdio.h>

printf("Hello from me!\n");
*/

/*
- If we compile the above, we will get an error.
- This is because the compiler was confused.
- Just like the Python interpreter expects correctly written single lines of code in a script,
- the C compiler gcc expects correctly written C code that it was programmed to translate.
- Though Python seemed so straightforward, it has its own language rules.
- C has more rules. This is because writing C code is more close to creating binary code than Python.
- For many of the rules, you will see that they map directly to the visualization of how binary code executes.
- This is why I think learning C is important, because it helps you understand binary structure of software.

- Rule 1: every C program must have a 'main()' function.
- The compiler looks for this function when preparing the binary code.
- It places this function in the final binary code such that execution begins with the code of this function.
- In this function, we can make calls to other functions, such as printf.
- The source of printf still has to be included. 

- Rule 2: every C function must have a return type
- A function computes something and can sometimes return the result of computation.
- In C, the type of this data must be specified when defining a new function.
- The printf function only displays to the screen, it does not return any data.
- A non-return type is called void.
- The main function must always return an integer. Reason for later.

*/

/*
#include<stdio.h>

int main() {
	printf("Hello from me!\n");

	return 0;
}
*/

/*
	compiling and running
*/

/*
	data types: string is char array
	passing multiple parameters to printf function
		noting string concatenation not a basic operation in C
	format strings: specify the type of data to interpret the parameter as
*/

/*
#include <stdio.h>

char name[] = "Yara";
char greeting[] = "Hello from me: ";

int main() {

	printf("%s %s\n", greeting, name);
	printf("%p %p\n", &greeting, &name);

	return 0;
}
*/

int num_1 = 5;
int num_2 = 7;

int sum() {
	int sum = num_1 + num_2;
	return sum;
}

//int sum()(int x, int y) {
//	int sum = x + y;
//	return sum;
//}

int main() {
	int my_sum = sum();
	//int my_sum = sum(num_1, num_2);
	printf("sum = %d\n", my_sum);
	return 0;
}

// NOTE data types, data sizes, all visible with objdump

/*
	from objdump, note the hex values that represent our strings
	goto data representation
*/

File: intro/python/primer.py
## Learning Objectives:
#
#
#
# Code = data + instructions
## A code file has variables and functions. Functions can use variables.
## Our code file is written in Python. It uses a 'print' function that prints a 'string' variable. A string is a sequence of ASCII characters.
## A variable 'greeting' can be used to store a string. Then, the print function can print the value of this variable.
#
#
# Execution = operation of written code
## A Python code file is converted to a runnable binary code by the
## python3 interpreter s/w. The interpreter parses the code line by 
## line and converts it to binary encodings.
## For a binary code to run, it is copied to memory.
## In the binary code, each function has a unique address/id,
## and each variable has a unique address/id.
## When executing, the processor/cpu can locate different functions and
## variables it needs through their unique memory addresses.
#
#
# Defining Variables and Functions
## Just like the print function that is defined in the Python language
## we can define our own functions.
## A 'greet' function can be defined to always use the Python 'print'
## function to display a greeting.
## During interpretation, user defined variable 'greeting' is first
## translated to its binary encoding: a region of the full binary code
## that contains the value of 'greeting'.
## Then, function 'greet' is translated to its binary encoding: a 
## region of the full binary that contains the instructions of the
## function.
## Finally, the call to 'greet' is converted to its binary encoding,
## which tells the processor to execute the instructions of 'greet'.



# PL constructs:
## making function calls - PL vocabulary
## function arguments
## constant data, data types

## try 0
## TODO: making a call to predefined print function, passing a
##	 string variable
## NOTE: python3 interpretter translates python code to binary code;
##	 the call to the print function becomes a sequence of binary
##	 encodings, and the characters of the greeting are translated
##	 to a binary representation of them;
##	 the OS then runs this code to print to screen.

print("Hello from me! Welcome to CS Summer Challenge :)")


# PL constructs:
## user-defined variables
## function argument passing

## try 1
## TODO: using variables
## NOTE: python3 translates variables to regions in the binary code
## 	 that hold chunks of data; it then translates the print
##	 function such that it knows the location of these chunks
##	 in the code
 
greeting = "Hello from me! Welcome to CS Summer Challenge :)"
print(greeting)

print(id(greeting))
print(id(print))


# PL constructs:
## basic operations

## try 2
## TODO: string concatenation
## NOTE: python3 translates the '+' operator to binary encodings that 
##	 concatenate the 3 different strings together to become the
##	 value of the greeting variable

name = "Yara"
greeting = "Hello from " + name + " ! Welcome to CS Summer Challenge :)"
print(greeting)

print(id(name))
print(id(greeting))
print(id(print))


# PL constructs:
## defining functions
## using variables inside defined functions

## try 3
## TODO: user-defined greet()
## NOTE: a custom function can be defined to be used; python3
##	 translates the function to a chunk of binary code in the
##	 full binary; this code knows the location of the 
##	 variable 'greeting' in the full binary
#
name = "Yara"
greeting = "Hello from " + name + " ! Welcome to CS Summer Challenge :)"

def greet():
	print(greeting)

greet()

print(id(name))
print(id(greeting))
print(id(print))
print(id(greet))


# PL constructs:
## program organization

## try 4
## NOTE: function must first be defined and interpreted, then it can be called

#name = "Yara"
#greeting = "Hello from " + name + " ! Welcome to CS Summer Challenge :)"
#
#greet()
#
#def greet():
#	print(greeting)


# PL constructs:
## defining a local variable inside a function
## returning value of local variable from a function

## try 5
## NOTE: understand using local variables; when translated to binary, the function
##	 binary encoding will have space reserved for a local variable, known only to
##	 this function, called sum; the value of this variable will be computed 
##	 by the function.
##	 understand returning values; returned values also have space reserved for them
##	 in the binary, however, they don't have names; hence, return values must be 
##	 captured by named variables if they are to be used - e.g. printed

num_1 = 5;
num_2 = 7;

def sum():
	sum = num_1 + num_2;
	return sum

my_sum = sum()
print("sum = ", my_sum)


# PL constructs:
## defining a function that expects parameters when it is called
## passing multiple parameters to a function

## try 6
## NOTE: understand parameter passing; when translated to binary, the function
##	 binary encoding will have space reserved for two parameters identified as
##	 x and y by the function; the code in the sum function will use the values
##	 of these two parameters 

num_1 = 5;
num_2 = 7;

def sum(x, y):
	sum = x + y;
	return sum

my_sum = sum(num_1, num_2)
print("sum = ", my_sum)



